// react-activity-calendar/build/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var chroma = require('chroma-js');
var dateFns = require('date-fns');
var React = require('react');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var chroma__default = /*#__PURE__*/_interopDefault(chroma);
var React__default = /*#__PURE__*/_interopDefault(React);

function _extends() {
  _extends = Object.assign ? Object.assign.bind() : function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];
      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }
    return target;
  };
  return _extends.apply(this, arguments);
}

const NAMESPACE = 'react-activity-calendar';
const LABEL_MARGIN = 8; // px

const DEFAULT_MONTH_LABELS = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
const DEFAULT_LABELS = {
  months: DEFAULT_MONTH_LABELS,
  weekdays: ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'],
  totalCount: '{{count}} activities in {{year}}',
  legend: {
    less: 'Less',
    more: 'More'
  }
};

function useColorScheme() {
  const [colorScheme, setColorScheme] = React.useState('light');
  const onChange = event => setColorScheme(event.matches ? 'dark' : 'light');
  React.useEffect(() => {
    const mediaQuery = window.matchMedia('(prefers-color-scheme: dark)');
    setColorScheme(mediaQuery.matches ? 'dark' : 'light');
    mediaQuery.addEventListener('change', onChange);
    return () => {
      mediaQuery.removeEventListener('change', onChange);
    };
  }, []);
  return colorScheme;
}

const query = '(prefers-reduced-motion: reduce)';
function usePrefersReducedMotion() {
  const [prefersReducedMotion, setPrefersReducedMotion] = React.useState(true);
  React.useEffect(() => {
    const mediaQuery = window.matchMedia(query);
    setPrefersReducedMotion(mediaQuery.matches);
    const onChange = event => {
      setPrefersReducedMotion(event.matches);
    };
    mediaQuery.addEventListener('change', onChange);
    return () => {
      mediaQuery.removeEventListener('change', onChange);
    };
  }, []);
  return prefersReducedMotion;
}

function styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') { return; }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css_248z = ".styles-module_container__GBNxC {\n  width: max-content; /* Calendar should not grow */\n  max-width: 100%; /* Do not remove - flexbox parents */\n  display: flex;\n  flex-direction: column;\n  gap: 8px;\n}\n\n.styles-module_container__GBNxC rect {\n  stroke: rgba(0, 0, 0, 0.08);\n  stroke-width: 1px;\n  shape-rendering: geometricPrecision;\n}\n\n@media (prefers-color-scheme: dark) {\n  .styles-module_container__GBNxC rect {\n    stroke: rgba(255, 255, 255, 0.04);\n  }\n}\n\n.styles-module_scrollContainer__-bJC8 {\n  max-width: 100%;\n  overflow-x: auto;\n  overflow-y: hidden;\n}\n\n.styles-module_calendar__sT1ND {\n  display: block; /* SVGs are inline-block by default */\n  overflow: visible; /* Weekday labels are rendered left of the container */\n}\n\n.styles-module_calendar__sT1ND text {\n  fill: currentColor;\n}\n\n.styles-module_footer__ZQ-Bl {\n  display: flex;\n  flex-wrap: wrap;\n  gap: 4px 16px;\n  white-space: nowrap;\n}\n\n.styles-module_legendColors__kFY2e {\n  margin-left: auto;\n  display: flex;\n  align-items: center;\n  gap: 3px;\n}\n\n@keyframes styles-module_loadingAnimation__V0w3h {\n  0% {\n    fill: var(--react-activity-calendar-loading);\n  }\n  50% {\n    fill: var(--react-activity-calendar-loading-active);\n  }\n  100% {\n    fill: var(--react-activity-calendar-loading);\n  }\n}\n";
var styles = {"container":"styles-module_container__GBNxC","scrollContainer":"styles-module_scrollContainer__-bJC8","calendar":"styles-module_calendar__sT1ND","footer":"styles-module_footer__ZQ-Bl","legendColors":"styles-module_legendColors__kFY2e","loadingAnimation":"styles-module_loadingAnimation__V0w3h"};
styleInject(css_248z);

function groupByWeeks(activities, weekStart = 0 // 0 = Sunday
) {
  if (activities.length === 0) {
    return [];
  }
  const normalizedActivities = fillHoles(activities);

  // Determine the first date of the calendar. If the first date is not the
  // set start weekday, the selected weekday one week earlier is used.
  const firstDate = dateFns.parseISO(normalizedActivities[0].date);
  const firstCalendarDate = dateFns.getDay(firstDate) === weekStart ? firstDate : dateFns.subWeeks(dateFns.nextDay(firstDate, weekStart), 1);

  // To correctly group activities by week, it is necessary to left-pad the list
  // because the first date might not be set start weekday.
  const paddedActivities = [...Array(dateFns.differenceInCalendarDays(firstDate, firstCalendarDate)).fill(undefined), ...normalizedActivities];
  const numberOfWeeks = Math.ceil(paddedActivities.length / 7);

  // Finally, group activities by week
  return Array(numberOfWeeks).fill(undefined).map((_, weekIndex) => paddedActivities.slice(weekIndex * 7, weekIndex * 7 + 7));
}

/**
 * The calendar expects a continuous sequence of days,
 * so fill gaps with empty activity data.
 */
function fillHoles(activities) {
  const dateMap = {};
  for (const activity of activities) {
    dateMap[activity.date] = activity;
  }
  return dateFns.eachDayOfInterval({
    start: dateFns.parseISO(activities[0].date),
    end: dateFns.parseISO(activities[activities.length - 1].date)
  }).map(day => {
    const date = dateFns.formatISO(day, {
      representation: 'date'
    });
    if (dateMap[date]) {
      return dateMap[date];
    }
    return {
      date,
      count: 0,
      level: 0
    };
  });
}
function getMonthLabels(weeks, monthNames = DEFAULT_MONTH_LABELS) {
  return weeks.reduce((labels, week, weekIndex) => {
    const firstActivity = week.find(activity => activity !== undefined);
    if (!firstActivity) {
      throw new Error(`Unexpected error: Week ${weekIndex + 1} is empty: [${week}].`);
    }
    const month = monthNames[dateFns.getMonth(dateFns.parseISO(firstActivity.date))];
    const prevLabel = labels[labels.length - 1];
    if (weekIndex === 0 || prevLabel.label !== month) {
      return [...labels, {
        weekIndex,
        label: month
      }];
    }
    return labels;
  }, []).filter(({
    weekIndex
  }, index, labels) => {
    // Labels should only be shown if there is "enough" space (data).
    // This is a naive implementation that does not take the block size,
    // font size etc. into account.
    const minWeeks = 3;

    // Skip the first month label if there is not enough space to the next one.
    if (index === 0) {
      return labels[1] && labels[1].weekIndex - weekIndex >= minWeeks;
    }

    // Skip the last month label the there is not enough data in that month to
    // avoid overflowing the calendar on the right.
    if (index === labels.length - 1) {
      return weeks.slice(weekIndex).length >= minWeeks;
    }
    return true;
  });
}
function getClassName(name, styles) {
  if (styles) {
    return `${NAMESPACE}__${name} ${styles}`;
  }
  return `${NAMESPACE}__${name}`;
}
function generateEmptyData() {
  const year = new Date().getFullYear();
  const days = dateFns.eachDayOfInterval({
    start: new Date(year, 0, 1),
    end: new Date(year, 11, 31)
  });
  return days.map(date => ({
    date: dateFns.formatISO(date, {
      representation: 'date'
    }),
    count: 0,
    level: 0
  }));
}
function maxWeekdayLabelLength(firstWeek, weekStart, labels, fontSize) {
  return firstWeek.reduce((maxLength, _, index) => {
    if (index % 2 !== 0) {
      const dayIndex = (index + weekStart) % 7;
      const curLength = Math.ceil(calcTextDimensions(labels[dayIndex], fontSize).width);
      return Math.max(maxLength, curLength);
    }
    return maxLength;
  }, 0);
}
function calcTextDimensions(text, fontSize) {
  if (typeof document === 'undefined' || typeof window === 'undefined') {
    throw new Error('calcTextDimensions() requires browser APIs');
  }
  if (fontSize < 1) {
    throw new RangeError('fontSize must be positive');
  }
  if (text.length === 0) {
    return {
      width: 0,
      height: 0
    };
  }
  const namespace = 'http://www.w3.org/2000/svg';
  const svg = document.createElementNS(namespace, 'svg');
  svg.style.position = 'absolute';
  svg.style.visibility = 'hidden';
  svg.style.fontFamily = window.getComputedStyle(document.body).fontFamily;
  svg.style.fontSize = `${fontSize}px`;
  const textNode = document.createElementNS(namespace, 'text');
  textNode.textContent = text;
  svg.appendChild(textNode);
  document.body.appendChild(svg);
  const boundingBox = textNode.getBBox();
  document.body.removeChild(svg);
  return {
    width: boundingBox.width,
    height: boundingBox.height
  };
}

function createTheme(input, size = 5) {
  const defaultTheme = createDefaultTheme(size);
  if (input) {
    validateTheme(input, size);
    input.light = input.light ?? defaultTheme.light;
    input.dark = input.dark ?? defaultTheme.dark;
    return {
      light: isColorScale(input.light, size) ? input.light : createColorScale(input.light, size),
      dark: isColorScale(input.dark, size) ? input.dark : createColorScale(input.dark, size)
    };
  }
  return defaultTheme;
}
function createDefaultTheme(size) {
  return {
    light: createColorScale(['hsl(0, 0%, 92%)', 'hsl(0, 0%, 26%)'], size),
    dark: createColorScale(['hsl(0, 0%, 20%)', 'hsl(0, 0%, 92%)'], size)
  };
}
function validateTheme(input, size) {
  if (typeof input !== 'object' || input.light === undefined && input.dark === undefined) {
    throw new Error(`The theme object must contain at least one of the fields "light" and "dark" with exactly 2 or ${size} colors respectively.`);
  }
  if (input.light) {
    const {
      length
    } = input.light;
    if (length !== 2 && length !== size) {
      throw new Error(`theme.light must contain exactly 2 or ${size} colors, ${length} passed.`);
    }
  }
  if (input.dark) {
    const {
      length
    } = input.dark;
    if (length !== 2 && length !== size) {
      throw new Error(`theme.dark must contain exactly 2 or ${size} colors, ${length} passed.`);
    }
  }
}
function isColorScale(colors, size) {
  const invalidColor = colors.find(color => !chroma__default.default.valid(color));
  if (invalidColor) {
    throw new Error(`Invalid color "${invalidColor}" passed. All CSS color formats are accepted.`);
  }
  return colors.length === size;
}
function createColorScale(colors, size) {
  return chroma__default.default.scale(colors).mode('lch').colors(size);
}

const ActivityCalendar = ({
  data,
  blockMargin = 4,
  blockRadius = 2,
  blockSize = 12,
  colorScheme = undefined,
  eventHandlers = {},
  fontSize = 14,
  hideColorLegend = false,
  hideMonthLabels = false,
  hideTotalCount = false,
  labels: labelsProp = undefined,
  maxLevel = 4,
  loading = false,
  renderBlock = undefined,
  showWeekdayLabels = false,
  style: styleProp = {},
  theme: themeProp = undefined,
  totalCount: totalCountProp = undefined,
  weekStart = 0 // Sunday
}) => {
  maxLevel = Math.max(1, maxLevel);
  const theme = createTheme(themeProp, maxLevel + 1);
  const systemColorScheme = useColorScheme();
  const colorScale = theme[colorScheme ?? systemColorScheme];
  const useAnimation = !usePrefersReducedMotion();
  if (loading) {
    data = generateEmptyData();
  }
  if (data.length === 0) {
    return null;
  }
  const year = dateFns.getYear(dateFns.parseISO(data[0]?.date));
  const weeks = groupByWeeks(data, weekStart);
  const labels = Object.assign({}, DEFAULT_LABELS, labelsProp);
  const labelHeight = hideMonthLabels ? 0 : fontSize + LABEL_MARGIN;
  const weekdayLabelOffset = showWeekdayLabels ? maxWeekdayLabelLength(weeks[0], weekStart, labels.weekdays, fontSize) + LABEL_MARGIN : undefined;
  function getDimensions() {
    return {
      width: weeks.length * (blockSize + blockMargin) - blockMargin,
      height: labelHeight + (blockSize + blockMargin) * 7 - blockMargin
    };
  }
  function getEventHandlers(activity) {
    return Object.keys(eventHandlers).reduce((handlers, key) => ({
      ...handlers,
      [key]: event => eventHandlers[key]?.(event)(activity)
    }), {});
  }
  function renderCalendar() {
    return weeks.map((week, weekIndex) => week.map((activity, dayIndex) => {
      if (!activity) {
        return null;
      }
      if (activity.level < 0 || activity.level > maxLevel) {
        throw new RangeError(`Provided activity level ${activity.level} for ${activity.date} is out of range. It must be between 0 and ${maxLevel}.`);
      }
      const style = loading && useAnimation ? {
        animation: `${styles.loadingAnimation} 1.75s ease-in-out infinite`,
        animationDelay: `${weekIndex * 20 + dayIndex * 20}ms`
      } : undefined;
      const block = /*#__PURE__*/React__default.default.createElement("rect", _extends({}, getEventHandlers(activity), {
        x: 0,
        y: labelHeight + (blockSize + blockMargin) * dayIndex,
        width: blockSize,
        height: blockSize,
        rx: blockRadius,
        ry: blockRadius,
        fill: colorScale[activity.level],
        "data-date": activity.date,
        "data-level": activity.level,
        style: style
      }));
      return /*#__PURE__*/React__default.default.createElement(React.Fragment, {
        key: activity.date
      }, renderBlock ? renderBlock(block, activity) : block);
    })).map((week, x) => /*#__PURE__*/React__default.default.createElement("g", {
      key: x,
      transform: `translate(${(blockSize + blockMargin) * x}, 0)`
    }, week));
  }
  function renderFooter() {
    if (hideTotalCount && hideColorLegend) {
      return null;
    }
    const totalCount = typeof totalCountProp === 'number' ? totalCountProp : data.reduce((sum, activity) => sum + activity.count, 0);
    return /*#__PURE__*/React__default.default.createElement("footer", {
      className: getClassName('footer', styles.footer),
      style: {
        marginLeft: weekdayLabelOffset
      }
    }, loading && /*#__PURE__*/React__default.default.createElement("div", null, "\xA0"), !loading && !hideTotalCount && /*#__PURE__*/React__default.default.createElement("div", {
      className: getClassName('count')
    }, labels.totalCount ? labels.totalCount.replace('{{count}}', String(totalCount)).replace('{{year}}', String(year)) : `${totalCount} activities in ${year}`), !loading && !hideColorLegend && /*#__PURE__*/React__default.default.createElement("div", {
      className: getClassName('legend-colors', styles.legendColors)
    }, /*#__PURE__*/React__default.default.createElement("span", {
      style: {
        marginRight: '0.4em'
      }
    }, labels?.legend?.less ?? 'Less'), Array(maxLevel + 1).fill(undefined).map((_, level) => /*#__PURE__*/React__default.default.createElement("svg", {
      width: blockSize,
      height: blockSize,
      key: level
    }, /*#__PURE__*/React__default.default.createElement("rect", {
      width: blockSize,
      height: blockSize,
      fill: colorScale[level],
      rx: blockRadius,
      ry: blockRadius
    }))), /*#__PURE__*/React__default.default.createElement("span", {
      style: {
        marginLeft: '0.4em'
      }
    }, labels?.legend?.more ?? 'More')));
  }
  function renderLabels() {
    if (!showWeekdayLabels && hideMonthLabels) {
      return null;
    }
    return /*#__PURE__*/React__default.default.createElement(React__default.default.Fragment, null, showWeekdayLabels && /*#__PURE__*/React__default.default.createElement("g", {
      className: getClassName('legend-weekday')
    }, weeks[0].map((_, index) => {
      if (index % 2 === 0) {
        return null;
      }
      const dayIndex = (index + weekStart) % 7;
      return /*#__PURE__*/React__default.default.createElement("text", {
        x: -LABEL_MARGIN,
        y: labelHeight + (blockSize + blockMargin) * index + blockSize / 2,
        dominantBaseline: "middle",
        textAnchor: "end",
        key: index
      }, labels.weekdays[dayIndex]);
    })), !hideMonthLabels && /*#__PURE__*/React__default.default.createElement("g", {
      className: getClassName('legend-month')
    }, getMonthLabels(weeks, labels.months).map(({
      label,
      weekIndex
    }) => /*#__PURE__*/React__default.default.createElement("text", {
      x: (blockSize + blockMargin) * weekIndex,
      dominantBaseline: "hanging",
      key: weekIndex
    }, label))));
  }
  const {
    width,
    height
  } = getDimensions();
  const containerStyles = {
    fontSize,
    ...(useAnimation && {
      [`--${NAMESPACE}-loading`]: colorScale[0],
      [`--${NAMESPACE}-loading-active`]: colorScheme === 'light' ? chroma__default.default(colorScale[0]).darken(0.3).hex() : chroma__default.default(colorScale[0]).brighten(0.25).hex()
    })
  };
  return /*#__PURE__*/React__default.default.createElement("article", {
    className: `${NAMESPACE} ${styles.container}`,
    style: {
      ...styleProp,
      ...containerStyles
    }
  }, /*#__PURE__*/React__default.default.createElement("div", {
    className: getClassName('scroll-container', styles.scrollContainer)
  }, /*#__PURE__*/React__default.default.createElement("svg", {
    width: width,
    height: height,
    viewBox: `0 0 ${width} ${height}`,
    className: getClassName('calendar', styles.calendar),
    style: {
      marginLeft: weekdayLabelOffset
    }
  }, !loading && renderLabels(), renderCalendar())), renderFooter());
};
const Skeleton = props => /*#__PURE__*/React__default.default.createElement(ActivityCalendar, _extends({
  data: []
}, props));

exports.Skeleton = Skeleton;
exports.default = ActivityCalendar;
